h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -776,182 +795,127 @@ private void readStoreHeader() {
-                            // used here as is, re-point our chunk to original
-                            // location instead.
-                            c.block = test.block;
-                            continue;
-                        }
-
-                        if (!c.isLive()) {
-                            // we can just remove entry from meta, referencing to this chunk,
-                            // but store maybe R/O, and it's not properly started yet,
-                            // so lets make this chunk "dead" and taking no space,
-                            // and it will be automatically removed later.
-                            c.block = Long.MAX_VALUE;
-                            c.len = Integer.MAX_VALUE;
-                            if (c.unused == 0) {
-                                c.unused = creationTime;
-                            }
-                            if (c.unusedAtVersion == 0) {
-                                c.unusedAtVersion = INITIAL_VERSION;
+            vdiff = newest.version - headerVersion;
+            lastCandidates[(int)vdiff] = newest;
+            newest = test;
+        }
+        
+        // In most cases, the newest is the last chunk, otherwise the store is corrupted by
+        // H2 bug, a rogue thread or process, even re-order file system.
+        // It's unnecessary that verifying the all chunk validity in the meta of last chunk candidates:
+        // 1) If the store corrupted by a rogue thread or process and we fall back to the older chunk,
+        // then the data lost, it's very bad. We should suggest recovering the store by tool to user or 
+        // in recovery mode.
+        // 2) And if the newest chunk valid(both header and footer valid), the all chunks referenced 
+        // in the meta of the newest also valid, otherwise why is the newest valid?, or the store corrupted.
+        // 3) The bug of database corruption by partial write and the store not closed when fatal error
+        // occurs in H2 has been fixed.
+        // @since 2019-08-08 little-pan
+        int i = lastCandidates.length;
+        do {
+            setLastChunk(newest);
+            // load the all chunk meta-data from the meta map in the last chunk and check recent 20 chunks
+            final Cursor<String, String> cursor = meta.cursor(DataUtils.META_CHUNK);
+            while (cursor.hasNext() && cursor.next().startsWith(DataUtils.META_CHUNK)) {
+                final Chunk c = Chunk.fromString(cursor.getValue());
+                assert c.version <= currentVersion;
+                // might be there already, due to meta traversal
+                // see readPage() ... getChunkIfFound()
+                chunks.putIfAbsent(c.id, c);
+                // only need to check the validity of recent chunks
+                vdiff = c.version - headerVersion;
+                if(vdiff < lastCandidates.length && vdiff >= 0L){
```

Comment:
Can we have spacing "if (...) {", which would be consistent with the rest of the code, here and elsewhere